lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,558 @@
+			Some(invoice) => self.pay_invoice_internal(invoice),
+			None => Err(PaymentError::Invoice("unknown payment hash")),
+		}
+	}
+
+	fn pay_invoice_internal(&self, invoice: &Invoice) -> Result<(), PaymentError> {
+		let payer = self.payer.node_id();
+		let payee = invoice.recover_payee_pub_key();
+		let payee_features = invoice.features().cloned();
+		let first_hops = self.payer.first_hops();
+		let last_hops = invoice.route_hints();
+		let final_value_msat = invoice.amount_pico_btc()
+			.ok_or(PaymentError::Invoice("amount missing"))? / 10;
+		let final_cltv = invoice.min_final_cltv_expiry() as u32;
+		let route = self.router.find_route(
+			&payer,
+			&payee,
+			payee_features,
+			Some(&first_hops.iter().collect::<Vec<_>>()),
+			&last_hops,
+			final_value_msat,
+			final_cltv,
+		).map_err(|e| PaymentError::Routing(e))?;
+
+		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+		let payment_secret = invoice.payment_secret().clone();
+		self.payer.send_payment(&route, payment_hash, &Some(payment_secret))
+			.map_err(|e| PaymentError::Sending(e))
+	}
+
+	/// Removes the [`Invoice`] cached by the given payment hash.
+	///
+	/// Should be called once a payment has failed or succeeded. This is taken care of when
+	/// [`InvoicePayer`] is used as an [`EventHandler`] but can be called independently as well.
+	pub fn remove_cached_invoice(&self, payment_hash: &PaymentHash) {
+		self.invoice_cache.lock().unwrap().remove(payment_hash);
+	}
+}
+
+impl<P: Deref, R, L: Deref, E> EventHandler for InvoicePayer<P, R, L, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	fn handle_event(&self, event: &Event) {
+		match event {
+			Event::PaymentFailed { payment_hash, rejected_by_dest, .. } => {
+				let mut attempts_by_payment_hash = self.payment_attempts.lock().unwrap();
+				let attempts = attempts_by_payment_hash
+					.entry(*payment_hash)
+					.and_modify(|attempts| *attempts += 1)
+					.or_insert(1);
+				if !rejected_by_dest {
+					let max_payment_attempts = self.retry_attempts + 1;
+					if *attempts < max_payment_attempts {
+						if self.pay_cached_invoice(payment_hash).is_ok() {
+							log_trace!(self.logger, "Payment {} failed; retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+							return;
+						} else {
+							log_trace!(self.logger, "Error retrying payment {}; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+						}
+					} else {
+						log_trace!(self.logger, "Payment {} failed; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+					}
+				} else {
+					log_trace!(self.logger, "Payment {} rejected by destination; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+				}
+
+				// Either the payment was rejected, exceeded the maximum attempts, or failed retry.
+				self.remove_cached_invoice(payment_hash);
+				attempts_by_payment_hash.remove(payment_hash);
```

Comment:
nit: it seems somewhat duplicative to get the entry above and then remove it here. You could just store the entry and remove the entry here, doing the and_modify.or_insert in the retry condition before the return.

Reply:
We need the number of attempts using `and_modify` and `or_insert` for the comparison, which consumes the entry. But if no retry is needed we, couldn't then remove the entry directly since it has been consumed.

Probably could just do `and_modify` then match to determine whether or not there's a value, but that doesn't seem as clean.

Reply:
Its not quite as clean, but it does feel incredibly wasteful to hash into the hashmap twice for the same element 5 lines apart :/

```
--- a/lightning-invoice/src/payment.rs
+++ b/lightning-invoice/src/payment.rs
@@ -202,11 +202,13 @@ where
                                let mut attempts_by_payment_hash = self.payment_attempts.lock().unwrap();
-                               let attempts = attempts_by_payment_hash
-                                       .entry(*payment_hash)
-                                       .and_modify(|attempts| *attempts += 1)
-                                       .or_insert(1);
+                               let attempts_entry = attempts_by_payment_hash
+                                       .entry(*payment_hash);
+                               let attempts = if let hash_map::Entry::Occupied(e) = &attempts_entry { e.get() + 1 } else { 1 };
                                if !rejected_by_dest {
                                        let max_payment_attempts = self.retry_attempts + 1;
-                                       if *attempts < max_payment_attempts {
+                                       if attempts < max_payment_attempts {
                                                if self.pay_cached_invoice(payment_hash).is_ok() {
                                                        log_trace!(self.logger, "Payment {} failed; retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
+                                                       attempts_entry
+                                                       .and_modify(|attempts| *attempts += 1)
+                                                       .or_insert(1);
                                                        return;
@@ -224,3 +226,2 @@ where
                                self.remove_cached_invoice(payment_hash);
-                               attempts_by_payment_hash.remove(payment_hash);
                        },
```

Reply:
Yeah, you still need to remove the entry later, but that does work. However, there's still an additional look-up when `remove_cached_invoice` is called. Given no tests failed when the attempts entry was not removed -- and that when called independently `remove_cached_invoice` won't remove the attempts entry -- it probably makes sense to combine these two hash maps to ensure consistency. Along with your improvement, that takes us to one look-up instead of three.